### DISCUSSION TITLE
LLaMA2 70B H100 性能問題

### DISCUSSION THREAD

**Discussion by ShinoharaHare** [hpcaitech/ColossalAI#5180] [Community | Q&A]
我用了 4 節點共 32 卡的 H100 機器測試了 LLaMA2 70B 的性能，但 TFLOPS 只有 170 左右，這似乎和 A100 差不多，我不太確定這是不是合理的一個數值，如果不是，想請教一下可能的問題點。

平行化參數如下：
- TP=8
- PP=1
- ZERO=2

```
num_samples: 3, dp_world_size: 4, flop_megatron: 6880279910154240.0, flop: 6780680423866368, avg_duration: 4.944727897644043, avg_throughput: 2.4268271679246647
Throughput: 2.43 samples/sec, TFLOPS per GPU by Megatron: 173.93, TFLOPS per GPU: 171.41
Max CUDA memory usage: 63.46 GB
```

另外，如果我要在 9 個 8 卡 H100 節點訓練 LLaMA2 70B，請問會建議用什麼樣的平行化配置？

**Comment by kurisusnowdeng** [comment]
显存似乎还有空间，是否试过提高batch size有什么结果。其次，是否确认了flash attention是被正常使用的。

**Reply by ShinoharaHare** [reply]
您好，感謝回覆！

- Batch Size=1
- Max Length=4096

沒開到 Flash Attention 2，不過有開到 SDPA，下面是開了 FA2 的結果，有快一點，但也是差不多
Batch Size 改 2 會 OOM，不管是 SPDA 或 FA2

```
num_samples: 3, dp_world_size: 4, flop_megatron: 6880279910154240.0, flop: 6780680423866368, avg_duration: 4.876276969909668, avg_throughput: 2.4608938487388334
Throughput: 2.46 samples/sec, TFLOPS per GPU by Megatron: 176.37, TFLOPS per GPU: 173.82
Max CUDA memory usage: 63.45 GB
```

**Reply by kurisusnowdeng** [reply]
@ShinoharaHare batch size只能开1挺反常的，我们会在类似场景下检测一下，感谢反馈

**Reply by ShinoharaHare** [reply]
@kurisusnowdeng 
我又用了單機 8 卡測試了一下 7B 模型，TFLOPS 高了許多，但奇怪的是開了 TP 後能用的 Batch Size 反而更小

- TP=1, BS=16
```
num_samples: 48, dp_world_size: 8, flop_megatron: 1.1976567844503552e+16, flop: 10598611339444224, avg_duration: 26.73107147216797, avg_throughput: 14.365305199224853
Throughput: 14.37 samples/sec, TFLOPS per GPU by Megatron: 448.04, TFLOPS per GPU: 396.49
Max CUDA memory usage: 70.93 GB
```

- TP=2, BS=12
```
num_samples: 36, dp_world_size: 4, flop_megatron: 8982425883377664.0, flop: 7948958504583168, avg_duration: 11.649053573608398, avg_throughput: 12.361519250476102
Throughput: 12.36 samples/sec, TFLOPS per GPU by Megatron: 385.54, TFLOPS per GPU: 341.18
Max CUDA memory usage: 56.90 GB
```
- TP=2, BS=16
```
OOM
```

- TP=4, BS=12
```
num_samples: 36, dp_world_size: 2, flop_megatron: 8982425883377664.0, flop: 7948958504583168, avg_duration: 7.082586765289307, avg_throughput: 10.165777333339708
Throughput: 10.17 samples/sec, TFLOPS per GPU by Megatron: 317.06, TFLOPS per GPU: 280.58
Max CUDA memory usage: 55.95 GB
```

- TP=4, BS=16
```
OOM
```

**Reply by flybird11111** [reply]
请问你在单卡测试7B模型分别是使用的哪一个plugin呀？

**Reply by ShinoharaHare** [reply]
這裡用的都是 `HybridParallelPlugin`

**Reply by ShinoharaHare** [reply]
@flybird11111 您好
我注意到最新的 commit 您似乎更新了 `DistCrossEntropy` 的功能，因此特別聲明一下先前的測試都是基於尚未更新 `DistCrossEntropy` 之前的原始碼

**Reply by flybird11111** [reply]
嗯嗯，好的

**Comment by flybird11111** [comment]
> @kurisusnowdeng 我又用了單機 8 卡測試了一下 7B 模型，TFLOPS 高了許多，但奇怪的是開了 TP 後能用的 Batch Size 反而更小
> 
> * TP=1, BS=16
> 
> ```
> num_samples: 48, dp_world_size: 8, flop_megatron: 1.1976567844503552e+16, flop: 10598611339444224, avg_duration: 26.73107147216797, avg_throughput: 14.365305199224853
> Throughput: 14.37 samples/sec, TFLOPS per GPU by Megatron: 448.04, TFLOPS per GPU: 396.49
> Max CUDA memory usage: 70.93 GB
> ```
> 
> * TP=2, BS=12
> 
> ```
> num_samples: 36, dp_world_size: 4, flop_megatron: 8982425883377664.0, flop: 7948958504583168, avg_duration: 11.649053573608398, avg_throughput: 12.361519250476102
> Throughput: 12.36 samples/sec, TFLOPS per GPU by Megatron: 385.54, TFLOPS per GPU: 341.18
> Max CUDA memory usage: 56.90 GB
> ```
> 
> * TP=2, BS=16
> 
> ```
> OOM
> ```
> 
> * TP=4, BS=12
> 
> ```
> num_samples: 36, dp_world_size: 2, flop_megatron: 8982425883377664.0, flop: 7948958504583168, avg_duration: 7.082586765289307, avg_throughput: 10.165777333339708
> Throughput: 10.17 samples/sec, TFLOPS per GPU by Megatron: 317.06, TFLOPS per GPU: 280.58
> Max CUDA memory usage: 55.95 GB
> ```
> 
> * TP=4, BS=16
> 
> ```
> OOM
> ```
感谢您，我会尽快查看和解决这个问题

**Reply by flybird11111** [reply]
torchrun --nnodes=1 --nproc_per_node=8 --rdzv_id=1 --rdzv_backend=c10d --rdzv_endpoint=localhost:12345 benchmark.py -x  -c 7b -g --max_length 4096 -p 3d -b 16 --tp=1 -s 1 --zero=1
num_samples: 0, dp_world_size: 8, flop_megatron: 0, flop: 0, avg_duration: 0.0, avg_throughput: 0.0
Throughput: 0.00 samples/sec, TFLOPS per GPU by Megatron: 0.00, TFLOPS per GPU: 0.00
Max CUDA memory usage: 66229.59 MB

torchrun --nnodes=1 --nproc_per_node=8 --rdzv_id=1 --rdzv_backend=c10d --rdzv_endpoint=localhost:12345 benchmark.py -x  -c 7b -g --max_length 4096 -p 3d -b 16 --tp=2 -s 1 --zero=1

num_samples: 0, dp_world_size: 4, flop_megatron: 0, flop: 0, avg_duration: 0.0, avg_throughput: 0.0
Throughput: 0.00 samples/sec, TFLOPS per GPU by Megatron: 0.00, TFLOPS per GPU: 0.00
Max CUDA memory usage: 59883.28 MB

您好，我测试下来是开了tp表现更好一点，同样也可以跑batch_size = 16.

**Reply by ShinoharaHare** [reply]
可以請您提供一下環境配置嗎？我可能排除一下版本差異再試試
另外想問一下，您的測試步數只有 1 所以數值都是 0，是否能提供有效的數值供我參考看看？

**Reply by ShinoharaHare** [reply]
@flybird11111 
請問一下您的顯存是多大的？
我這裡的 80 GB，在 ZERO=2, TP=1, BS=16 基本就極限了，ZERO 改 1 就 OOM 了
環境是 colossalai==0.3.4, torch==2.0.1+cuda11.8

另外，我發現新的版本 colossalai==https://github.com/hpcaitech/ColossalAI/commit/79718fae04fc4461a35ae80ab87f52b64260f394, torch==2.1.1+cuda12.1 用跟上面一樣的配置跑不起來，跑了 1 步就 OOM

**Reply by flybird11111** [reply]
好的，我正在看。

**Reply by flybird11111** [reply]
因为开了tp，dp size就会减小, (tp size * dp size = world size)，可能batchsize在达到这个数值时，是增大tp size的收益不如增大dp size.

**Reply by ShinoharaHare** [reply]
速度上收益降低是挺合理，但理論上顯存不是該用的更少？
我看 OOM 都是出在 Cross Entropy 的部分，可能您實作的 `DistCrossEntropy` 可以解決這個問題？
雖然我沒有很確定這個東西的用途

**Reply by flybird11111** [reply]
DistCrossEntropy确实可以优化内存，因为增加dp size减小了，会导致dp group每个rank的batchsize增大，所以会出现oom

**Reply by flybird11111** [reply]
torchrun --nnodes=1 --nproc_per_node=8 --rdzv_id=1 --rdzv_backend=c10d --rdzv_endpoint=localhost:12345 benchmark.py -x  -c 7b -g --max_length 4096 -p 3d -b 16 --tp=2 -s 10 --zero=1
num_samples: 144, dp_world_size: 4, flop_megatron: 3.5929703533510656e+16, flop: 31795834018332672, avg_duration: 135.9298858642578, avg_throughput: 4.23747872910892
Throughput: 4.24 samples/sec, TFLOPS per GPU by Megatron: 132.16, TFLOPS per GPU: 116.96
Max CUDA memory usage: 50226.26 MB 我这边可以跑起来诶，我用A800跑的

**Reply by ShinoharaHare** [reply]
我的理解是 DP 所分散的優化器狀態與梯度比起 TP 分散模型參數更有效率，因為這裡的 Batch Size 不是本來就是指 Batch Size per Rank，這樣應該跟 DP Size 無關才對

不過我測 7B 主要是想和 70B 對比看看 TFLOPS，主要還是想解決 70B 的問題，想請問一下跨節點 70B 的場景下是不是也有可能把 TFLOPS 提高到 300~400 的水平？

**Reply by ShinoharaHare** [reply]
那真奇怪了，會不會是版本問題

**Reply by flybird11111** [reply]
不好意思，我未使用DistCrossEntropy,增加step也oom了,但是使用了DistCrossEntropy是可以运行的。

**Reply by flybird11111** [reply]
您节点之间的通信好吗？

**Reply by ShinoharaHare** [reply]
規格是寫 InfiniBand 100 Gbps，實際上沒測試過，應該算不差嗎？這部分我不是很懂

**Reply by ShinoharaHare** [reply]
目前用 6 節點，ZERO=2, TP=4, BS=2 可以達到 270 TFLOPS，請問有什麼優化建議嗎？

```
num_samples: 6, dp_world_size: 12, flop_megatron: 1.376055982030848e+16, flop: 13561360847732736, avg_duration: 12.496347427368164, avg_throughput: 5.76168359742524
Throughput: 5.76 samples/sec, TFLOPS per GPU by Megatron: 275.29, TFLOPS per GPU: 271.31
Max CUDA memory usage: 60.15 GB
```

**Reply by flybird11111** [reply]
目前这个设置是最优的吗？您可以尝试增大tp？我们测试的实际通信带宽在68GBps左右。

**Reply by ShinoharaHare** [reply]
是的，TP=8, BS=4 會稍微慢一點，變 264 TFLOPS 左右

**Reply by ShinoharaHare** [reply]
@flybird11111
請問您 GBps 的 B 應該是指小寫的 b bit 嗎？
我使用 iperf3 測試了一下各節點間的速度，有的可以到 34 Gbit/s，有的則是 20 Gbit/s，看來瓶頸就是出在這？

**Reply by flybird11111** [reply]
GBps是大写的B，没有写错.

**Comment by ShinoharaHare** [comment]
您好，我在訓練 LLaMA2 70B 的時候，存檔和讀檔有遇到 OOM 的問題，因此想另外請教一下

配置：
ColossalAI: 79718fae04fc4461a35ae80ab87f52b64260f394
Plugin: `HybridParallelPlugin`
- zero_stage=2
- tp_size=4
- pp_size=1
- precision=bf16
Hardware: 6 (Nodes) x 8 H100
Batch Size: 2
Max Length: 4096

狀況：
1. 存檔後 OOM：訓練正常運行，前幾次的存檔也正常，大概在第 5 次成功存檔後，訓練就 CUDA OOM 了，有沒有可能存檔的過程有造成顯存洩漏？

```python3
self.booster.save_model(
    self.boosted_model,
    os.path.join(checkpoint_path, _CKPT_MODEL_DIR),
    shard=True,
    size_per_shard=1024,
    use_safetensors=True
)
self.booster.save_optimizer(
    self.optimizers[0],
    os.path.join(checkpoint_path, _CKPT_OPTIMIZER_DIR),
    shard=True,
    size_per_shard=1024
)
```

2. 載入優化器狀態 OOM：直接就 CUDA OOM 了，根據下面的原始碼片段來說，是不是因為載入的時候是先載入完整的優化器狀態才分片導致的？理論上不是應該邊載入邊分片才能省顯存
https://github.com/hpcaitech/ColossalAI/blob/79718fae04fc4461a35ae80ab87f52b64260f394/colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py#L572-L606

```python3
self.booster.load_model(self.boosted_model, os.path.join(checkpoint_path, _CKPT_MODEL_DIR))
self.booster.load_optimizer(self.optimizers[0], os.path.join(checkpoint_path, _CKPT_OPTIMIZER_DIR))
```

**Reply by flybird11111** [reply]
不存档会出现oom吗

**Reply by ShinoharaHare** [reply]
應該是不會，因為我確定是存完檔再繼續 forward 的時候立刻就 OOM 了，應該是跟存檔有直接關係沒錯
不過後續會再確定一下

**Reply by ShinoharaHare** [reply]
@flybird11111 您好，後續測試了一下，確實不存檔就不會 OOM

**Reply by flybird11111** [reply]
感谢，我看看

**Reply by ShinoharaHare** [reply]
@flybird11111
存擋和讀檔問題有後續嗎？
不知道 GeminiPlugin 有沒有相同問題，我後來試 Gemini + offload 速度似乎比 Hybrid + TP 更快，如果它沒有這個問題，我可能就先用 Gemini 了

**Reply by flybird11111** [reply]
可以的，gemini保存和加载是没有问题的。

**Comment by flybird11111** [comment]
> * zero_stage

<img width="774" alt="image" src="https://github.com/hpcaitech/ColossalAI/assets/37931082/3e31a2e4-ed6e-48ab-854a-e970238f34ba">
<img width="1060" alt="image" src="https://github.com/hpcaitech/ColossalAI/assets/37931082/6c3c0033-3b92-4a41-9845-e6e1d3479aa0">

Hi，看起来Hybrid ParallelPlugin存档也没有影响显存的使用。

> @flybird11111 存擋和讀檔問題有後續嗎？ 不知道 GeminiPlugin 有沒有相同問題，我後來試 Gemini + offload 速度似乎比 Hybrid + TP 更快，如果它沒有這個問題，我可能就先用 Gemini 了
